// view/BaseInstrument.js
var inherit = require("../utils/inherit");
var BaseView = require("./BaseView");
var commonEventNames = require("../core/commonEventNames");
var windowsTransport = require("./windowsTransport");

module.exports = BaseInstrument;

function BaseInstrument(track, instrumentName){
    BaseView.call(this, instrumentName);
    this.track = track;
    this.startTime = 0;
    this.isRecordNow = false;
}

inherit(BaseInstrument, BaseView);

BaseInstrument.prototype.recordEvent = function(eventName, recordButton){
    if(this.isRecordNow === true){
        // stop record
        this.isRecordNow = false;
        recordButton.text("record");
        recordButton.attr("data-content", "Click to record sound");
        windowsTransport.notify(commonEventNames.E_UPDATE_WAVE);
    } else{
        this.isRecordNow = true;
        recordButton.text("stop");
        recordButton.attr("data-content", "Click to stop record");
        this.clearEvent();
    }
};

BaseInstrument.prototype.clearEvent = function(){
    if(this.track){
        console.log("clear recorded");
        this.track.emptyPlaySetting(); // clear previous play data setting
    }
};

BaseInstrument.prototype.keyDown = function(){};

BaseInstrument.prototype.keyUp = function(){};
