// view/CoordMap.js
module.exports = CoordMap;

function CoordMap(){
    this.map = $("<div class='coord-map'>");
}

CoordMap.prototype.getY = function(y){
    var newCoord = 0;
    var newHeight = getEven(this.map.height());
    var center = newHeight / 2;
    if(y === center){
        // begin coord
        newCoord = 0;
    } else if(y === 0){
        // max coord
        newCoord = center;
    } else if(y < center && y > 0){
        // positive
        // new coord for positive part is just division of center and y
        newCoord = getNewCoord(y, center);
    } else if(y > center && y <= newHeight){
        // convert to negative
        newCoord = (center - getNewCoord(y, newHeight)) * (-1);
    }
    return newCoord;
};

CoordMap.prototype.getX = function(x){
    var newCoord = 0;
    var newWidth = getEven(this.map.width());
    var center = newWidth / 2;
    if(x === center){
        // begin coord
        newCoord = 0;
    } else if(x === 0){
        // max coord
        newCoord = newWidth / -2;
    } else if(x < center && x > 0){
        // convert to negative
        newCoord = getNewCoord(x, center) * (-1);
    } else if(x > center && x <= newWidth){
        // positive
        newCoord = center - getNewCoord(x, newWidth);
        console.log(newCoord);
    }
    return newCoord;
};

function getNewCoord(coord, limit){
    return limit - coord
}

function getEven(number){
    var evenNumber = number;
    if(number % 2 !== 0){
        evenNumber = number - 1;
    }
    return evenNumber;
}
