// view/Factory.js
module.exports = {
    "createKey": createKey,
    "createInput": createInput,
    "createButton": createButton,
    "rangeElement": rangeElement,
    "buttonsPopup": buttonsPopup,
    "setColorToKey": setColorToKey,
    "createCheckBox": createCheckBox,
    "dropDownElement": dropDownElement,
    "createIconButton": createIconButton,
    "createButtonPopup": createButtonPopup,
    "deleteCircleButton": deleteCircleButton,
    "setBeginValueToRangeElement": setBeginValueToRangeElement
};

function createButton(className, text){
    return $("<button class='ui button " + className + "'>" + text + "</button>");
}

function createIconButton(buttonClass, iconClass, name){
    return $("<button class='" + buttonClass + "'>" +
        "<i class='" + iconClass + "'></i>" + name + "</button>");
}

function createButtonPopup(buttonClass, iconClass, name, message){
    var $button;
    if(iconClass){
        $button = $("<button class='" + buttonClass + "' data-content='" + message + "'>" +
                    "<i class='" + iconClass + "'></i>" + name + "</button>");
    } else{
        $button = $("<button class='ui button " + buttonClass + "' data-content='" + message + "'>"
                    + name + "</button>");
    }
    $button.popup();
    return $button;
}

function deleteCircleButton(buttonId, callback){
    var $deleteProjectButton = createIconButton("circular ui icon button delete", "remove icon", "");
    $deleteProjectButton.attr("id", buttonId);
    $deleteProjectButton.on("click", function(){
        callback($(this));
    });
    return $deleteProjectButton;
}

function createInput(className, text){
    var $result = $("<div class='ui input " + className + "'>");
    var $input = $("<input type='text'>");
    $input.value = text;
    $result.append($input);
    return $result;
}

/**
 * Create range element (equalizer) with label with value of it
 * @param className
 * @param id - String - element name
 * @param minValue
 * @param maxValue
 * @param callback
 * @param stepValue
 * @param beginValue
 * @returns {*|jQuery|HTMLElement}
 */
function rangeElement(className, id, minValue, maxValue, callback, stepValue, beginValue){
    var value;
    var step = stepValue || 1;
    var begin = beginValue || 0;
    var $result = $("<div class='" + className + "'>");
    var $label = $("<div class='ui label'>" + begin + "</div>");
    var $element = $("<input id='" + id + "' type='range' step=" + step + " min=" + minValue +
                     " max=" + maxValue + " value=" + begin + ">");
    $element.on("input", function(){
        value = $(this).val();
        $label.text(value);
        callback(id, Number(value));
    });
    $result.append($element);
    $result.append($label);
    return $result;
}

/**
 * Set value to range element and its label
 * @param rangeElement
 * @param id
 * @param beginValue
 */
function setBeginValueToRangeElement(rangeElement, id, beginValue){
    rangeElement.find('#' + id).value = beginValue; // set input range value
    rangeElement.find(".ui.label").text(beginValue); // set label value
}

/**
 *
 * @param className
 * @param id
 * @param text
 * @param checkCallback
 * @param uncheckCallback
 * @param isActive
 * @returns {*|jQuery|HTMLElement}
 */
function createCheckBox(className, id, text, checkCallback, uncheckCallback, isActive){
    var checked = isActive || false;
    var $result = $("<div class='ui checkbox " + className + "'>");
    var $input = $("<input id='" + id + "' type='checkbox'>");
    var $text = $("<label>" + text + "</label>");
    $input.prop("checked", checked);
    $input.change(function(){
        if($(this).prop("checked") === true){
            checkCallback($(this).attr("id"), $(this).prop("checked"));
        } else{
            uncheckCallback($(this).attr("id"), $(this).prop("checked"));
        }
    });
    $result.append($input);
    $result.append($text);
    return $result;
}

/**
 *
 * @param className
 * @param name
 * @param dataObject - Object - key-element name, value-data-value
 * @param callback
 * @param defaultValue
 * @returns {string}
 */
function dropDownElement(className, name, dataObject, callback, defaultValue){
    var key;
    var value = defaultValue || 0;
    var $item;
    var $menu = $("<div class='menu'>");
    var $dropdown = $("<div class='ui selection dropdown'>");
    var $result = $("<div class='" + className + "'></div>");
    $dropdown.append("<div class='default text'>default</div>");
    $dropdown.append("<input name='choice' value='" + value + "' type='hidden'>");
    $dropdown.append("<i class='dropdown icon'>");
    for(key in dataObject){
        $item = $("<div id='" + name + "' class='item' data-value='" + dataObject[key] + "'>" + key + "</div>");
        $item.on("click", function(){
            callback($(this).attr("id"), $(this).text(), $(this).attr("data-value"));
        });
        $menu.append($item);
    }
    $dropdown.append($menu);
    $dropdown.dropdown();
    $result.append($dropdown);
    return $result;
}

/**
 * Create button set with the same callback
 * @param buttonNameList
 * @param callback
 */
function buttonsPopup(buttonNameList, callback){
    var i;
    var $column;
    var $button;
    var $result = $("<div class='ui flowing popup top left transition hidden'>");
    var $row = $("<div class='ui two column divided center aligned grid'>");
    for(i = 0; i < buttonNameList.length; ++i){
        $column = $("<div class='column'>");
        $button = createButton("", buttonNameList[i]);
        $button.on("click", function(event){
            callback($(this).text());
        });
        $column.append($button);
        $row.append($column);
    }
    $result.append($row);
    return $result;
}

function keyContent(key, note){
    var $result = $("<div class='key-content'>");
    var $note = $("<label class='note'>" + note + "</label>");
    var $key = $("<label class='keyboard " + key + "'>" + key + "</label>");
    $result.append($note);
    $result.append("<br>");
    $result.append($key);
    return $result;
}

var isClicked = false;
function createKey(className, id, style, key, note, mouseDownCallback, mouseUpCallback){
    var $key = $("<div class='" + className + "' id='" + id + "' style='" + style + "'>");
    $key.append(keyContent(key, note));
    $key.mousedown(function(event){
        isClicked = true;
        mouseDownCallback(key);
    });
    $key.mouseup(function(event){
        isClicked = false;
        mouseUpCallback(key);
    });
    $key.mouseenter(function(event){
        if(isClicked === true){
            mouseDownCallback(key);
        }
    });
    $key.mouseleave(function(event){
        mouseUpCallback(key);
    });
    return $key;
}

function setColorToKey(pianoElement, key, color){
    pianoElement.find('#' + key).css("background-color", color);
}
