// view/UserModal.js
var inherit = require("../utils/inherit");
var Factory = require("./Factory");
var BaseView = require("./BaseView");
var commonEventNames = require("../core/commonEventNames");
var windowsTransport = require("./windowsTransport");

module.exports = UserModal;

function UserModal(){
    BaseView.call(this, "ui modal user-registration");
    this.buttons = $("<div class='buttons-registration'>");
    this.facebookLogIn = Factory.createIconButton("ui facebook button", "facebook icon", "Facebook");
    this.googleLogIn = Factory.createIconButton("ui google plus button", "google plus icon", "Google");
    //this.label = $("<div class='ui pointing below label'>Join without registration");
    this.buttonLogout = Factory.createButton("", "Logout");

    this._build();
}

inherit(UserModal, BaseView);

UserModal.prototype._build = function(){
    this.facebookLogIn.on("click", function(event){
        location.href = "/accounts/facebook/login/";
    });

    this.googleLogIn.on("click", function(event){
        location.href = "/accounts/google/login/";
    });

    this.buttonLogout.on("click", function(event){
        location.href = "/accounts/logout/";
    });

    windowsTransport.subscribe(commonEventNames.E_SHOW_LOGIN_FORM, this.show.bind(this));

    this.buttons.append(this.facebookLogIn);
    this.buttons.append(this.googleLogIn);
    this.buttons.append(this.buttonLogout);
    this.getContainer().append(this.buttons);
};

UserModal.prototype.show = function(){
    this._container.modal("show");
};
