// view/WaveForm.js
var inherit = require("../utils/inherit");
var AudioHelper = require("../utils/AudioHelper");
var BaseView = require("./BaseView");
var commonEventNames = require("../core/commonEventNames");
var windowsTransport = require("./windowsTransport");

module.exports = WaveForm;

function WaveForm(){
    BaseView.call(this, "wave-form");

    this.wavesurfer = null;

    this._build();
}

inherit(WaveForm, BaseView);

WaveForm.prototype._build = function(){
};

WaveForm.prototype.create = function(buffer){
    var self = this;
    this.wavesurfer = WaveSurfer.create({
        container: this.getContainer().get(0),
        waveColor: "#262626",
        progressColor: "#ebebeb",
        height: 120
    });
    this.wavesurfer.on("ready", function(){
        self.wavesurfer.drawer.container.style.display = '';
        self.wavesurfer.drawBuffer();
    });
    this.wavesurfer.loadBlob(AudioHelper.AudioBufferToBlob(buffer));
    windowsTransport.subscribe(commonEventNames.E_PLAY_WAVE, function(){
        self.wavesurfer.setMute(true);
        self.wavesurfer.stop();
        self.wavesurfer.play();
    });
    windowsTransport.subscribe(commonEventNames.E_STOP_WAVE, this.wavesurfer.stop.bind(this.wavesurfer));
};
