// model/AudioPlayer.js
var BaseTrackModel = require("./BaseTrackModel");
var ProjectModel = require("./ProjectModel");
var commonEventNames = require("../core/commonEventNames");
var windowsTransport = require("../view/windowsTransport");

module.exports = AudioPlayer;

function AudioPlayer(model){
    this.model = null;

    this.projectPlayer = null;
    this.recorder = null;

    this.setModel(model);
}

AudioPlayer.prototype.setModel = function(model){
    if(model){
        console.log("new model");
        console.log(model);
        this.model = model;
    }
};

AudioPlayer.prototype.play = function(){
    if(this.model instanceof BaseTrackModel){
        console.log("play");
        console.log(this.model);
        this.model.play();
        windowsTransport.notify(commonEventNames.E_PLAY_WAVE);
    } else if(this.model instanceof ProjectModel){
        console.log("? ? ?");
        windowsTransport.notify(commonEventNames.E_PLAY_PROJECT, this);
    }
};

AudioPlayer.prototype.createPlayer = function(buffer){
    var self = this;
    this.projectPlayer = new Tone.Player(buffer).toMaster();
    this.projectPlayer.start();
    windowsTransport.subscribe(commonEventNames.E_STOP_PROJECT, function(){
        self.projectPlayer.stop();
    });
};

AudioPlayer.prototype.stop = function(){
    if(this.model instanceof BaseTrackModel){
        this.model.stop();
    } else if(this.model instanceof ProjectModel){
        windowsTransport.notify(commonEventNames.E_STOP_PROJECT);
    }
    windowsTransport.notify(commonEventNames.E_STOP_WAVE);
};
