// model/BaseTrackModel.js
var PostSettings = require("./PostSettings");
var generateUID = require("../core/generateUID");
var AudioHelper = require("../utils/AudioHelper");

module.exports = BaseTrackModel;

// @param {String} name
// @param {Object} source
function BaseTrackModel(id, data){
    this.isDeleted = false;
    this.id = id || -generateUID();
    this.length = data.length || 1;
    this.setting = data.setting || {}; // this
    this.instrument = data.instrument || "synth";
    this.playSetting = data["play-setting"] || [];
    this.playObjects = [];

    this.trackObject = this._generate();
    this.postSettings = new PostSettings(data["post-setting"]);

    this.createPlayObjects();
}

BaseTrackModel.prototype.createFormData = function(){
    var formData = new FormData();
    formData.append("user_audio", toBlob(this.trackObject), this.name);
    formData.append("postProcessSettings", JSON.stringify(this.postProcessSettings));
    return formData;
};

BaseTrackModel.prototype.getContext = function(){
    return this.trackObject.context._context;
};

BaseTrackModel.prototype.getConstants = function(){
    return this.trackObject.context;
};

BaseTrackModel.prototype.getAudioBuffer = function(){
    return AudioHelper.getAudioContextBuffer(this.getConstants());
};

BaseTrackModel.prototype.getBlob = function(){
    return AudioHelper.AudioContextToBlob(this.getConstants());
};

BaseTrackModel.prototype.getVolume = function(){
    return this.trackObject ? this.trackObject.volume.value : null;
};

BaseTrackModel.prototype.getType = function(){
    return this.trackObject ? this.trackObject.type : null;
};

/**
 * Return object of track setting in right format for transfer or saving
 * @returns {{}}
 */
BaseTrackModel.prototype.getData = function(){
    var result = {};
    result.id = this.id;
    result.isDeleted = this.isDeleted;
    result.instrument = this.instrument;
    result.length = this.length;
    result.setting = this.setting;
    result["play-setting"] = this.playSetting;
    result["post-setting"] = this.postSettings.getPostSettings();
    return result;
};

BaseTrackModel.prototype.toJson = function(){
    return JSON.stringify(this);
};

BaseTrackModel.prototype.setVolume = function(value){
    this.trackObject.volume.value = value;
};

BaseTrackModel.prototype.setType = function(value){
    this.trackObject.type = value;
};

/**
 * Set data to play setting list from play setting object
 */
BaseTrackModel.prototype.setPlaySettings = function(){
    var i;
    this.playSetting.length = 0; // clear previous play setting data
    for(i = 0; i < this.playObjects.length; ++i){
        this.playSetting.push(this.playObjects[i].getData());
    }
};

BaseTrackModel.prototype.emptyPlaySetting = function(){
    this.playSetting.length = 0;
    this.playObjects.length = 0;
};

BaseTrackModel.prototype.disconnectFromAudioSource = function(){
    this.trackObject.disconnect(Tone.Master);
};

BaseTrackModel.prototype.clearObject = function(){
    this.trackObject = null;
};

/**
 * @param {Object} source
 * @type {null}
 * @protected
 * @return {Tone}
 */
BaseTrackModel.prototype._generate = null;

BaseTrackModel.prototype.createPlayObjects = null;

/**
 * @param {Object} options
 * @type {null}
 */
BaseTrackModel.prototype.play = null;

BaseTrackModel.prototype.record = null;

BaseTrackModel.prototype.setSetting = null;

BaseTrackModel.prototype.applyFilter = null;

BaseTrackModel.prototype.removeFilter = null;

BaseTrackModel.prototype.merge = null;
