// model/MixerRecorder.js
var inherit = require("../utils/inherit");
var clamp = require("../utils/clamp");
var BaseRecorder = require("./BaseRecorder");

module.exports = MixerRecorder;


function MixerRecorder(frequency, volume, startTime, stopTime){
    BaseRecorder.call(this);
    this.frequency = frequency;
    this.volume = volume;
    this.startTime = startTime || 0;
    //this.stopTime = stopTime || 0;
    this.checkValues();
}

inherit(MixerRecorder, BaseRecorder);

MixerRecorder.prototype.checkValues = function(){
    this.frequency = clamp(this.frequency, 0, 1500);
    this.volume = clamp(this.volume, -75, 75);
};

MixerRecorder.prototype.play = function(oscillator){
    var self = this;
    setTimeout(function(){
        oscillator.frequency.value = self.frequency;
        oscillator.volume.value = self.volume;
    }, this.startTime);
    oscillator.start('+' + (this.startTime / 1000));
    //oscillator.stop(this.stopTime / 1000);
};

MixerRecorder.prototype.getData = function(){
    var result = {};
    result.frequency = this.frequency;
    result.volume = this.volume;
    result.startTime = this.startTime;
    result.stopTime = this.stopTime;
    return result;
};
