// model/PianoKeyRecorder.js
var inherit = require("../utils/inherit");
var BaseRecorder = require("./BaseRecorder");

module.exports = PianoKeyRecorder;

/**
 * @param piano - Tone.Synth
 * @param note - String
 * @param triggerAttackTime
 * @param triggerReleaseTime
 * @constructor
 */
function PianoKeyRecorder(piano, note, triggerAttackTime, triggerReleaseTime){
    BaseRecorder.call(this);
    this.piano = piano;
    this.note = note;
    this.triggerAttackTime = triggerAttackTime || 0; // milliseconds
    this.triggerReleaseTime = triggerReleaseTime || 0; // milliseconds
}

inherit(PianoKeyRecorder, BaseRecorder);

PianoKeyRecorder.prototype.play = function(){
    // convert to seconds:
    var start = this.triggerAttackTime / 1000;
    var duration = (this.triggerReleaseTime - this.triggerAttackTime) / 1000;
    this.piano.triggerAttackRelease(this.note, duration, '+' + start);
};

PianoKeyRecorder.prototype.getData = function(){
    var result = {};
    result.note = this.note;
    result.triggerAttackTime = this.triggerAttackTime;
    result.triggerReleaseTime = this.triggerReleaseTime;
    return result;
};
