// model/PostSettings.js
var FreeverbFilter = require("./FreeverbFilter");
var CrusherFilter = require("./CrusherFilter");
var PhaserFilter = require("./PhaserFilter");
var VibratoFilter = require("./VibratoFilter");
var TremoloFilter = require("./TremoloFilter");

module.exports = PostSettings;

/**
 * Contain set of FilterModel objects.
 * Wait object with filters from server.
 * Or if track was created from the client set empty object for future settings.
 * @param postSettings - object
 * @constructor
 */
function PostSettings(postSettings){
    this.postSettings = postSettings || {};
    this.filterObjects = {};
    this.fullFilterObjects();
}

/**
 * Create corresponding filter object
 * @param filterName
 * @param filterSetting
 */
PostSettings.prototype.createFilterObjects = function(filterName, filterSetting){
    switch(filterName){
        case "tremolo":
            this.filterObjects[filterName] = new TremoloFilter(filterSetting);
            break;
        case "vibrato":
            this.filterObjects[filterName] = new VibratoFilter(filterSetting);
            break;
        case "crusher":
            this.filterObjects[filterName] = new CrusherFilter(filterSetting);
            break;
        case "phaser":
            this.filterObjects[filterName] = new PhaserFilter(filterSetting);
            break;
        case "freeverb":
            this.filterObjects[filterName] = new FreeverbFilter(filterSetting);
            break;
    }
};

PostSettings.prototype.getFilterModel = function(filterName){
    var result = null;
    if(filterName in this.filterObjects){
        result = this.filterObjects[filterName];
    }
    return result;
};

/**
 * Apply filter to track from track model
 * @param filterName - String
 * @param trackModel - BaseTrackModel
 */
PostSettings.prototype.setToTrack = function(filterName, trackModel){
    var filter = this.getFilterModel(filterName);
    if(filter !== null){
        //trackModel.disconnectFromAudioSource();
        if(trackModel.applyFilter === null) {
            filter.applyToTrack(trackModel.trackObject);
        } else{
            trackModel.applyFilter(filter);
        }
    }
};

PostSettings.prototype.removeFromTrack = function(filterName, trackModel){
    var filter = this.getFilterModel(filterName);
    if(filter !== null){
        if(trackModel.removeFilter === null) {
            filter.disconnectFilter(trackModel.trackObject);
        } else{
            trackModel.removeFilter(filter);
        }
    }
};

/**
 * Create list of filter objects with settings from server.
 * Call when init PostSettings.
 */
PostSettings.prototype.fullFilterObjects = function(){
    var filterName;
    for(filterName in this.postSettings){
        this.createFilterObjects(filterName, this.postSettings[filterName]);
    }
};

/**
 * Return object with applying filters only
 * @returns {{}}
 */
PostSettings.prototype.getPostSettings = function(){
    var filter;
    var filterName;
    var result = {};
    for(filterName in this.filterObjects){
        filter = this.filterObjects[filterName];
        result[filterName] = filter.getOptions();
        /*if(filter.isUsed === true){
            // save only used filters
            result[filterName] = filter.getOptions();
        }*/
    }
    return result;
};

/**
 * Set new filter settings in corresponding filter object and switch isUsed flag
 * @param filterList - TrackSettingsSet - FilterList
 */
PostSettings.prototype.setPostSettings = function(filterList){
    var i;
    var tokenSetting;
    var list = filterList.list;
    for(i = 0; i < list.length; ++i){
        tokenSetting = list[i];
        this.setFilter(tokenSetting);
    }
};

/**
 * Add or delete filter from filterObjects
 * @param filter
 */
PostSettings.prototype.setFilter = function(filter){
    var name = filter.name;
    if(name in this.filterObjects){
        if(filter.isEnabled === true){
            // update existing filter
            this.filterObjects[name].setOptions(filter.valueOf());
        } else{
            //delete no using filter in track
            delete this.filterObjects[name];
        }
    } else{
        //create new filter
        if(filter.isEnabled === true){
            this.createFilterObjects(name, filter.valueOf());
        }
    }
};

PostSettings.prototype.setValueToFilter = function(filterName, optionName, value){
    var filter;
    if(filterName in this.filterObjects){
        filter = this.filterObjects[filterName];
        if(optionName in filter.options){
            filter.setByName(optionName, value);
        }
    }
};
