// model/TrackOscillator.js
var BaseTrackModel = require("./BaseTrackModel");
var inherit = require("../utils/inherit");
var MixerRecorder = require("./MixerRecorder");

module.exports = TrackOscillator;

// @param {String} name
// @param {Object} source
function TrackOscillator(id, data){
    BaseTrackModel.call(this, id, data);
}

inherit(TrackOscillator, BaseTrackModel);

TrackOscillator.prototype._generate = function(){
    return new Tone.Oscillator(this.setting).toMaster();
};

TrackOscillator.prototype.getFrequency = function(){
    return this.trackObject ? this.trackObject.frequency.value : null;
};

TrackOscillator.prototype.setFrequency = function(value){
    this.trackObject.frequency.value = value;
};

TrackOscillator.prototype.createPlayObjects = function(){
    var i, tokenPlaySetting;
    for(i = 0; i < this.playSetting.length; ++i){
        tokenPlaySetting = this.playSetting[i];
        this.playObjects.push(new MixerRecorder(tokenPlaySetting.frequency, tokenPlaySetting.volume,
                                                tokenPlaySetting.startTime));
    }
};

TrackOscillator.prototype.play = function(){
    var i;
    for(i =0; i < this.playObjects.length; ++i){
        this.playObjects[i].play(this.trackObject);
    }
    this.trackObject.stop();
};

TrackOscillator.prototype.stop = function(){
    this.trackObject.stop();
};

TrackOscillator.prototype.setSetting = function(){
    if(this.trackObject){
        if(this.setting === undefined){
            this.setting = {};
        }
        this.setting.type = this.getType();
    }
};
