// model/TrackSynthesizer.js
var BaseTrackModel = require("./BaseTrackModel");
var PianoKeyPlayer = require("./PianoKeyRecorder");
var inherit = require("../utils/inherit");

module.exports = TrackSynthesizer;

// @param {String} name
// @param {Object} source
function TrackSynthesizer(id, data){
    BaseTrackModel.call(this, id, data);
}

inherit(TrackSynthesizer, BaseTrackModel);

TrackSynthesizer.prototype.getType = function(){
    return this.trackObject.oscillator.type;
};

TrackSynthesizer.prototype.getAttack = function(){
    return this.trackObject.envelope.attack;
};

TrackSynthesizer.prototype.getDecay = function(){
    return this.trackObject.envelope.decay;
};

TrackSynthesizer.prototype.getSustain = function(){
    return this.trackObject.envelope.sustain;
};

TrackSynthesizer.prototype.getRelease = function(){
    return this.trackObject.envelope.release;
};

TrackSynthesizer.prototype.setType = function(value){
    this.trackObject.oscillator.type = value;
};

TrackSynthesizer.prototype.setAttack = function(value){
    this.trackObject.envelope.attack = value;
};

TrackSynthesizer.prototype.setDecay = function(value){
    this.trackObject.envelope.decay = value;
};

TrackSynthesizer.prototype.setSustain = function(value){
    this.trackObject.envelope.sustain = value;
};

TrackSynthesizer.prototype.setRelease = function(value){
    this.trackObject.envelope.release = value;
};

TrackSynthesizer.prototype.setSetting = function(){
    if(this.trackObject){
        if(this.setting.oscillator === undefined){
            this.setting.oscillator = {};
        }
        if(this.setting.envelope === undefined){
            this.setting.envelope = {};
        }
        this.setting.oscillator.valume = this.getVolume();
        //this.setting.oscillator.frequency = this.getFrequency();
        //this.setting.frequency = this.getFrequency();
        this.setting.oscillator.type = this.getType();
        this.setting.envelope.attack = this.getAttack();
        this.setting.envelope.decay = this.getDecay();
        this.setting.envelope.sustain = this.getSustain();
        this.setting.envelope.release = this.getRelease();
    }
};

TrackSynthesizer.prototype._generate = function(){
    return new Tone.Synth(this.setting).toMaster(); //{"oscillator": {}, "envelope": {}}
};

TrackSynthesizer.prototype.play = function(){
    var i;
    for(i = 0; i < this.playObjects.length; ++i){
        this.playObjects[i].play();
    }
};

TrackSynthesizer.prototype.stop = function(){
    this.trackObject.triggerRelease();
};

TrackSynthesizer.prototype.playKeyNow = function(note){
    try {
        this.trackObject.triggerAttack(note);
    } catch(SyntaxError){
        console.log(note);
    }
};

TrackSynthesizer.prototype.stopKeyNow = function(){
    this.trackObject.triggerRelease();
};

TrackSynthesizer.prototype.createPlayObjects = function(){
    var i, tokenPlaySetting;
    for(i = 0; i < this.playSetting.length; ++i){
        tokenPlaySetting = this.playSetting[i];
        this.playObjects.push(new PianoKeyPlayer(this.trackObject, tokenPlaySetting.note,
            tokenPlaySetting.triggerAttackTime, tokenPlaySetting.triggerReleaseTime));
    }
};
