// model/TremoloFilter.js
var inherit = require("../utils/inherit");
var BaseFilterModel = require("./BaseFilterModel");

module.exports = TremoloFilter;

function TremoloFilter(options){
    BaseFilterModel.call(this, options);
}

inherit(TremoloFilter, BaseFilterModel);

TremoloFilter.prototype.generate = function(){
    this.filter = new Tone.Tremolo(this.options);
};

TremoloFilter.prototype.getFrequency = function(){
    return this.filter.frequency.value;
};

TremoloFilter.prototype.getDepth = function(){
    return this.filter.depth.value;
};

TremoloFilter.prototype.getSpread = function(){
    return this.filter.spread;
};

TremoloFilter.prototype.getType = function(){
    return this.filter.type;
};

TremoloFilter.prototype.getOptions = function(){
    return this.options;
};

TremoloFilter.prototype.setFrequency = function(value){
    this.options.frequency = value;
    this.filter.frequency.value = value;
};

TremoloFilter.prototype.setDepth = function(value){
    this.options.depth = value;
    this.filter.depth.value = value;
};

TremoloFilter.prototype.setSpread = function(value){
    this.options.spread = value;
    this.filter.spread = value;
};

TremoloFilter.prototype.setType = function(value){
    this.options.type = value;
    this.filter.type = value;
};

TremoloFilter.prototype.setOptions = function(options){
    this.setFrequency(options.frequency);
    this.setDepth(options.depth);
    this.setSpread(options.spread);
    this.setType(options.type);
};

TremoloFilter.prototype.setByName = function(optionName, value){
    switch(optionName){
        case "frequency":
            this.setFrequency(value);
            break;
        case "depth":
            this.setDepth(value);
            break;
        case "spread":
            this.setSpread(value);
            break;
        case "type":
            this.setType(value);
            break;
    }
};

TremoloFilter.prototype.applyToTrack = function(track){
    this.filter.toMaster();
    //track.disconnect(Tone.Master);
    track.connect(this.filter);
    this.filter.start();
};

TremoloFilter.prototype.disconnectFilter = function(track){
    track.disconnect(this.filter);
    //track.toMaster();
    this.filter.stop();
};
